### PR TITLE
Add basic documentation of the CSRF validator

### PR DESCRIPTION
**Pull request by settermjd**
Signed-off-by: Matthew Setter <matthew@matthewsetter.com>

|    Q          |   A
|-------------- | ------
| Documentation | yes
| Bugfix        | no
| BC Break      | no
| New Feature   | no
| RFC           | no
| QA            | no

### Description

This PR updates the documentation to add core documentation of the CSRF validator. 
It includes the essentials of:

- An introductory description
- The additional required packages
- The supported options
- A basic usage example

### CODE REVIEW

File: docs/book/validators/csrf.md [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+`Laminas\Validator\Csrf` provides the ability to both generate and validate CSRF tokens.
+This allows you to validate if a form submission originated from the same site, by confirming the value of the CSRF field in the submitted form is the same as the one contained in the original form.
+
+[Cross-Site Request Forgery (CSRF)](https://en.wikipedia.org/wiki/Cross-site_request_forgery) is a security vector in which an unauthorized request is accepted by a server on behalf of another user; it is essentially an exploit of the trust a site places on a user's browser.
+
+The typical mitigation is to create a one-time token that is transmitted as part of the original form, and which must then be transmitted back by the client.
+This token expires after first submission or after a short amount of time, preventing replays or further submissions.
+If the token provided does not match what was originally sent, an error should be returned.
+
+## Supported options
```
Comment (froschdesign) [inline docs/book/validators/csrf.md]: For headlines we use the Chicago Manual of Style capitalization rules.

```suggestion
## Supported Options
```

File: docs/book/validators/csrf.md [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+| `timeout` | The [TTL](https://en.wikipedia.org/wiki/Time_to_live) for the CSRF token | Optional |
+
+## Library requirements
+
+Before you can use this validator, you have to install the following, additional, packages:
+
+- [laminas-math](https://docs.laminas.dev/laminas-math/)
+- [laminas-session](https://docs.laminas.dev/laminas-session/)
+
+## Basic usages
```
Comment (froschdesign) [inline docs/book/validators/csrf.md]:
```suggestion
## Basic Usage
```

File: docs/book/validators/csrf.md [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+Here is a basic, working, example.
```
Comment (froschdesign) [inline docs/book/validators/csrf.md]: "Working"? Some parts of the example are missing, therefore I would avoid this paragraph.

File: docs/book/validators/csrf.md [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+]);
+$hash = $validator->getHash();
+
+// ...Render the hash in the form.
+
+// Validate the hash after form submission.
+echo ($validator->isValid($hash))
+    ? "Token is valid"
+    : "Token is NOT valid";
+```
```
Comment (froschdesign) [inline docs/book/validators/csrf.md]: The target group of the documentation are experienced PHP programmers, so we can omit some lines.

```suggestion
```php
// Initialise a new session container
// or use the existing one in your application
$session = new Laminas\Session\Container();

// Create the validator
$validator = new Laminas\Validator\Csrf([
    'session' => $session,
]);
$hash = $validator->getHash();

// ...Render the hash in the form.

// Validate the hash after form submission.
echo $validator->isValid($hash)
    ? 'Token is valid'
    : 'Token is NOT valid';
```
```

File: docs/book/validators/csrf.md [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+| `salt` | The salt for the CSRF token | Optional |
+| `session` | The name of the session element containing the CSRF element | **Mandatory** |
+| `timeout` | The [TTL](https://en.wikipedia.org/wiki/Time_to_live) for the CSRF token | Optional |
+
+## Library requirements
+
+Before you can use this validator, you have to install the following, additional, packages:
+
+- [laminas-math](https://docs.laminas.dev/laminas-math/)
+- [laminas-session](https://docs.laminas.dev/laminas-session/)
```
Comment (froschdesign) [inline docs/book/validators/csrf.md]:
```suggestion
```

File: docs/book/validators/csrf.md [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+`Laminas\Validator\Csrf` provides the ability to both generate and validate CSRF tokens.
+This allows you to validate if a form submission originated from the same site, by confirming the value of the CSRF field in the submitted form is the same as the one contained in the original form.
+
+[Cross-Site Request Forgery (CSRF)](https://en.wikipedia.org/wiki/Cross-site_request_forgery) is a security vector in which an unauthorized request is accepted by a server on behalf of another user; it is essentially an exploit of the trust a site places on a user's browser.
+
+The typical mitigation is to create a one-time token that is transmitted as part of the original form, and which must then be transmitted back by the client.
+This token expires after first submission or after a short amount of time, preventing replays or further submissions.
+If the token provided does not match what was originally sent, an error should be returned.
+
```
Comment (froschdesign) [inline docs/book/validators/csrf.md:11]: >
> ### Installation Requirements
>
> The CSRF validator depends on the laminas-math component to generate the hash and on laminas-session to transport the token from one request to another, so both components must be installed:
>
> ```bash
> $ composer require laminas/laminas-math laminas/laminas-session
> ```
>

**Review by froschdesign — CHANGES_REQUESTED** [review]
Only a few small changes are necessary. Thanks in advance! 👍

**Comment by froschdesign** [comment]
@settermjd 
Please use the `2.14.x` branch for this pull request.

File: docs/book/validators/csrf.md (hunk 3 above)
Comment (settermjd) [reply docs/book/validators/csrf.md]: Really? Technically, it does work. It's a rather simplistic example, however.

File: docs/book/validators/csrf.md [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+| Option | Description | Optional/Mandatory |
+|-|-|-|
+| `name` | The name of the CSRF element | Optional |
+| `salt` | The salt for the CSRF token | Optional |
+| `session` | The name of the session element containing the CSRF element | **Mandatory** |
+| `timeout` | The [TTL](https://en.wikipedia.org/wiki/Time_to_live) for the CSRF token | Optional |
+
+## Library requirements
+
+Before you can use this validator, you have to install the following, additional, packages:
```
Comment (glensc) [inline docs/book/validators/csrf.md]: feels like superfluous commas around "additional".

File: docs/book/validators/csrf.md (hunk 3 above)
Comment (froschdesign) [reply docs/book/validators/csrf.md]: > It's a rather simplistic example, however.

Correct and this good here. But under "working example" I would personally expect something more, like "it works in my application".

File: docs/book/validators/csrf.md (hunk 7 above)
Comment (boesing) [reply docs/book/validators/csrf.md]: Nah, when I translate it to german, its like an extra hint that these packages are additional. Not necessarily superfluous tho. 
I guess that Matthew, as a native speaker and technical writer, knows best 😅

File: docs/book/validators/csrf.md (hunk 3 above)
Comment (settermjd) [reply docs/book/validators/csrf.md]: Understood.

File: docs/book/validators/csrf.md (hunk 7 above)
Comment (settermjd) [reply docs/book/validators/csrf.md]: I'm a bit of a comma-lover. I can't deny it. The comma may be superfluous, yet adds a little extra specificity to the sentence.

File: docs/book/validators/csrf.md [hunk 8]
```diff
@@ -9,6 +9,16 @@ The typical mitigation is to create a one-time token that is transmitted as part
...
 
+## Installation Requirements
+
+The CSRF validator depends on [laminas-math](https://docs.laminas.dev/laminas-math/) to generate the hash and on [laminas-session](https://docs.laminas.dev/laminas-session/) to persist the generated token between requests.
+Consequently, both components must be installed before the validator can be used.
+To install both components, run the following command.
+
+```bash
+composer require laminas/laminas-math laminas/laminas-session
+```
```
Comment (froschdesign) [inline docs/book/validators/csrf.md:20]: We use a blockquote for this. Please compare with:

* https://docs.laminas.dev/laminas-validator/validators/db/
* https://docs.laminas.dev/laminas-validator/validators/sitemap/#loc
* https://docs.laminas.dev/laminas-i18n/validators/introduction/
* https://docs.laminas.dev/laminas-paginator/usage/#rendering-pages-with-view-scripts
* …

```suggestion
> ### Installation Requirements
>
>The CSRF validator depends on [laminas-math](https://docs.laminas.dev/laminas-math/) to generate the hash and on [laminas-session](https://docs.laminas.dev/laminas-session/) to persist the generated token between requests.
> Consequently, both components must be installed before the validator can be used.
> To install both components, run the following command.
>
> ```bash
> composer require laminas/laminas-math laminas/laminas-session
> ```
```

**Comment by settermjd** [comment]
Closing in favour of #100.

**Pull request closed without merging**